daemon/containerd/image_snapshot_unix.go

```diff
@@ -0,0 +1,70 @@
+//go:build !windows
+
+package containerd
+
+import (
+	"context"
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+
+	"github.com/containerd/containerd/leases"
+	"github.com/containerd/containerd/mount"
+	"github.com/containerd/containerd/snapshots"
+	"github.com/docker/docker/pkg/idtools"
+)
+
+func (i *ImageService) remapSnapshot(ctx context.Context, snapshotter snapshots.Snapshotter, id string, parentSnapshot string, lease leases.Lease) error {
+	ls := i.client.LeasesService()
+	rootPair := i.idMapping.RootPair()
+	usernsID := fmt.Sprintf("%s-%d-%d", parentSnapshot, rootPair.UID, rootPair.GID)
+	remappedID := usernsID + remapSuffix
+
+	// If the remapped snapshot already exist we only need to prepare the new snapshot
+	if _, err := snapshotter.Stat(ctx, usernsID); err == nil {
+		_, err = snapshotter.Prepare(ctx, id, usernsID)
+		return err
+	}
+
+	if err := ls.AddResource(ctx, lease, leases.Resource{
+		ID:   remappedID,
+		Type: "snapshots/" + i.StorageDriver(),
+	}); err != nil {
+		return err
+	}
+	if err := ls.AddResource(ctx, lease, leases.Resource{
+		ID:   usernsID,
+		Type: "snapshots/" + i.StorageDriver(),
+	}); err != nil {
+		return err
+	}
+
+	mounts, err := snapshotter.Prepare(ctx, remappedID, parentSnapshot)
+	if err != nil {
+		return err
+	}
+
+	if err := remapRootFS(ctx, mounts, rootPair); err != nil {
```

Comment:
Why does this only use the rootPair and not the full mapping?

Reply:
`RootPair` is the remapped root uid/gid

Reply:
Comparing to containerd's implementation which increments over the full userns range https://github.com/containerd/containerd/blob/v1.7.5/container_opts_unix.go#L104

Reply:
Ah right!

Reply:
Fixed thanks